src/main/java/net/minecraftforge/common/crafting/NBTIngredient.java

```diff
@@ -26,17 +26,32 @@
 import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
 
+import net.minecraft.advancements.critereon.NbtPredicate;
+import net.minecraft.util.GsonHelper;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.crafting.Ingredient;
 import net.minecraft.network.FriendlyByteBuf;
 
 public class NBTIngredient extends Ingredient
 {
     private final ItemStack stack;
-    protected NBTIngredient(ItemStack stack)
+    private final boolean exactMatch;
+    private final NbtPredicate predicate;
+    protected NBTIngredient(ItemStack stack, boolean exactMatch)
     {
         super(Stream.of(new Ingredient.ItemValue(stack)));
         this.stack = stack;
+        this.exactMatch = exactMatch;
+
+        if (this.exactMatch)
+            this.predicate = NbtPredicate.ANY;
```

Comment:
I understand that this predicate is never actually used if exact match is true, but this does feel weird setting it to any when it is a very specific thing it expects.

Reply:
One option would be setting it to null, and using a null check on the predicate instead of the boolean field to determine whether to use it.

Reply:
I do agree it is weird, I didn't feel comforting putting it as `null`. If you have any other suggestions or think it should be null, I'm all ears,

Reply:
Ah, good point Knight. I'm a little `null`-phobic these days but that works